### PR TITLE
Upgrade log4j to 2.15.0

### PR DESCRIPTION
**Pull request by arteam**
Originally we tried to a log4j update in #47298, but we were unable to that due to the `DeprecationLoggerTests.testLogPermissions` test failing. The test relied on mocking and got removed as part of refactoring in
#61474.

Now we should be able to the upgrade and then we can address the Security Manager permission questions raised in #47298 separately.

### CODE REVIEW

File: plugins/discovery-gce/build.gradle [hunk 1]
```diff
@@ -52,6 +52,12 @@ tasks.named("thirdPartyAudit").configure {
           'org.apache.avalon.framework.logger.Logger',
           'org.apache.log.Hierarchy',
           'org.apache.log.Logger',
+          'org.apache.avalon.framework.logger.Logger',
```
Comment (pgomulka) [inline plugins/discovery-gce/build.gradle:55]: I guess we only need jms.Message? the rest are already declared?

File: plugins/repository-gcs/build.gradle [hunk 2]
```diff
@@ -180,7 +182,7 @@ tasks.named("thirdPartyAudit").configure {
     'org.apache.http.protocol.HttpRequestExecutor',
     // commons-logging provided dependencies
     'javax.servlet.ServletContextEvent',
-    'javax.servlet.ServletContextListener'
+    'javax.servlet.ServletContextListener',
```
Comment (pgomulka) [inline plugins/repository-gcs/build.gradle:185]: comma not needed?

File: plugins/repository-s3/build.gradle [hunk 3]
```diff
@@ -308,6 +309,6 @@ tasks.named("thirdPartyAudit").configure {
           'com.amazonaws.services.kms.model.EncryptResult',
           'com.amazonaws.services.kms.model.GenerateDataKeyRequest',
           'com.amazonaws.services.kms.model.GenerateDataKeyResult',
-          'javax.activation.DataHandler'
+          'javax.activation.DataHandler',
```
Comment (pgomulka) [inline plugins/repository-s3/build.gradle:312]: is this comma needed?

File: x-pack/plugin/core/build.gradle [hunk 4]
```diff
@@ -118,7 +118,15 @@ tasks.named("thirdPartyAudit").configure {
           'org.apache.log.Logger',
           //commons-logging provided dependencies
           'javax.servlet.ServletContextEvent',
-          'javax.servlet.ServletContextListener'
+          'javax.servlet.ServletContextListener',
```
Comment (pgomulka) [inline x-pack/plugin/core/build.gradle:121]: I think some entries are duplicated?

**Review by pgomulka — CHANGES_REQUESTED** [review]
left few questions

File: plugins/discovery-gce/build.gradle (hunk 1 above)
Comment (arteam) [reply plugins/discovery-gce/build.gradle:55]: Let me check that

File: plugins/repository-gcs/build.gradle (hunk 2 above)
Comment (arteam) [reply plugins/repository-gcs/build.gradle:185]: Got automatically added by IntelliJ, fixing

File: plugins/repository-s3/build.gradle (hunk 3 above)
Comment (arteam) [reply plugins/repository-s3/build.gradle:312]: Fixing

File: plugins/discovery-gce/build.gradle (hunk 1 above)
Comment (arteam) [reply plugins/discovery-gce/build.gradle:55]: You're right! We should add only one `javax.jms.Message` entry here

File: x-pack/plugin/core/build.gradle (hunk 4 above)
Comment (arteam) [reply x-pack/plugin/core/build.gradle:121]: You're right! I've removed the superfluous entries.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Review by pgomulka — APPROVED** [review]
LGTM, thank you !

**Comment by arteam** [comment]
@elasticmachine run  elasticsearch-ci/part-1

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by arteam** [comment]
@elasticmachine upgrade branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by wallrik** [comment]
When you work on this pull request, if you look at [Log4j 2.16.0](https://logging.apache.org/log4j/2.x/changes-report.html#a2.16.0) instead, it looks like the hack from #81629 can _probably_ be removed, if that is something you'd want to do.

**Comment by Xon** [comment]
> When you work on this pull request, if you look at [Log4j 2.16.0](https://logging.apache.org/log4j/2.x/changes-report.html#a2.16.0) instead, it looks like the hack from #81629 can _probably_ be removed, if that is something you'd want to do.

2.15.0 is affected by [CVE-2021-45046](https://github.com/advisories/GHSA-7rjr-3q55-vv33) to boot.

**Comment by mark-vieira** [comment]
FYI we should revert https://github.com/elastic/elasticsearch/pull/81629 as part of this as well as it's no longer necessary, especially if we go strait to 2.16 which disabled JNDI support entirely by default.

**Comment by Li4n0** [comment]
log4j-core has released version 2.16.0, which fixes a denial-of-service attack vulnerability in 2.15.0, would you consider upgrading directly to 2.16.0? @arteam

**Comment by arteam** [comment]
@Li4n0 We will do the 2.15.0 to 2.16.0 upgrade separately. Unfortunately, it's not a simple version bump, because 2.16.0 forbids things like `${sys:es.logs.cluster_name}` in pattern layouts which are used in Elasticsearch

**Comment by arteam** [comment]
Thanks @ChrisHegarty and @pgomulka!

**Pull request merged**